python_modules/dagster/dagster/_core/remote_representation/external_data.py

```diff
@@ -1560,213 +1552,128 @@ def external_asset_checks_from_defs(
+            compute_kind = node_def.tags.get("kind")
+            node_definition_name = node_def.name
+
+            # Confusingly, the `name` field sometimes mismatches the `name` field on the
+            # OutputDefinition. We need to fix this.
+            output_name = node_def.output_def_named(output_handle.output_name).name
+            required_top_level_resources = (
+                list(node_def.required_resource_keys) if isinstance(node_def, OpDefinition) else []
             )
-            if asset_key in metadata_by_asset_key
-            else output_def.metadata
-        )
-
-        job_names = [job_def.name for _, job_def in node_tuple_list]
-
-        partitions_def_data: Optional[ExternalPartitionsDefinitionData] = None
 
-        partitions_def = asset_info.partitions_def
-        if partitions_def:
-            partitions_def_data = external_partitions_definition_from_def(partitions_def)
+        else:
+            graph_name = None
+            op_names = []
+            op_name = None
+            job_names = []
+            compute_kind = None
+            node_definition_name = None
+            output_name = None
+            required_top_level_resources = []
+
+        # This is in place to preserve an implicit behavior in the Dagster UI where stub
+        # dependencies were rendered as if they weren't part of the default asset group.
+        group_name = None if asset_node.assets_def.is_auto_created_stub else asset_node.group_name
 
-        # if the asset is produced by an op at the top level of the graph, graph_name should be None
-        graph_name = None
-        node_handle = node_output_handle.node_handle
-        while node_handle.parent:
-            node_handle = node_handle.parent
-            graph_name = node_handle.name
         asset_nodes.append(
             ExternalAssetNode(
-                asset_key=asset_key,
-                dependencies=list(deps[asset_key].values()),
-                depended_by=list(dep_by[asset_key].values()),
-                execution_type=execution_types_by_asset_key[asset_key],
-                compute_kind=node_def.tags.get("kind"),
-                # backcompat
-                op_name=graph_name
-                or next(iter(op_names_by_asset_key[asset_key]), None)
-                or node_def.name,
+                asset_key=key,
+                dependencies=[
+                    ExternalAssetDependency(
+                        upstream_asset_key=pk,
+                        partition_mapping=(
+                            asset_graph.get_partition_mapping(key, pk)
+                            if asset_node.partitions_def and asset_graph.get(pk).partitions_def
+                            else None
+                        ),
+                    )
+                    for pk in sorted(asset_node.parent_keys)
+                ],
+                depended_by=[ExternalAssetDependedBy(k) for k in sorted(asset_node.child_keys)],
```

Comment:
Now that there is only one loop over all asset keys to construct `ExternalAssetNode`, all assets have dependencies and depended_by set correctly.